presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -746,7 +763,43 @@ public void alterTable(String databaseName, String tableName, Table table)
     }
 
     @Override
-    public void addPartitions(String databaseName, String tableName, List<Partition> partitions)
+    public void addPartitions(String databaseName, String tableName, List<PartitionWithStatistics> partitionsWithStatistics)
+    {
+        List<Partition> partitions = partitionsWithStatistics.stream()
+                .map(ThriftMetastoreUtil::toMetastoreApiPartition)
+                .collect(toImmutableList());
+        addMetastorePartitions(databaseName, tableName, partitions);
+        try {
+            if (partitionsWithStatistics.stream().allMatch(p -> p.getStatistics().getColumnStatistics().isEmpty())) {
+                return;
+            }
+            Table table = getTable(databaseName, tableName)
+                    .orElseThrow(() -> new TableNotFoundException(new SchemaTableName(databaseName, tableName)));
+            List<String> partitionColumns = table.getPartitionKeys().stream()
+                    .map(FieldSchema::getName)
+                    .collect(toImmutableList());
+            for (PartitionWithStatistics partitionWithStatistics : partitionsWithStatistics) {
+                String partitionName = makePartName(partitionColumns, partitionWithStatistics.getPartition().getValues());
+                storePartitionColumnStatistics(databaseName, tableName, partitionName, partitionWithStatistics);
+            }
+        }
+        catch (Throwable t) {
+            for (Partition partition : partitions) {
+                try {
+                    dropPartition(databaseName, tableName, partition.getValues(), false);
```

Comment:
Given there is `SemiTransactionalMetastore` several layers up, why is compensating actions necessary here?

Reply:
`add_partitions` method is supposed to be `atomic`. Please read the comment here: https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/HiveMetastore.java#L57. Othewise if the code storing statistics fails we will end up with created partitions. `SemiTransactionalMetastore` rolls back only successfully created partitions.